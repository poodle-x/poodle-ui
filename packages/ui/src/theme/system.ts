import {
	BordersProps,
	bordersSystem,
	ColorsProps,
	colorsSystem,
	compose,
	createSystem,
	FlexProps,
	flexSystem,
	FontsProps,
	fontsSystem,
	getMode,
	GridProps,
	gridSystem,
	LayoutProps,
	layoutSystem,
	ListProps,
	listSystem,
	PseudosProps,
	pseudosSystem,
	ResponsiveProp,
	ShadowProps,
	shadowSystem,
	SpacingProps,
	spacingSystem,
	SystemConfigList,
} from "@poodle/system";
import * as CSS from "csstype";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

// Cache to avoid unnecessary repeat loop
// @todo maybe move to a helper func
const finalSystem: SystemConfigList = {};

for (const keys in allSystem) {
	const keysPart = keys.split(",");

	const rawConfig = allSystem[keys];

	keysPart.forEach((k) => {
		finalSystem[k] = rawConfig;
	});
}

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};

	for (const prop in props) {
		let found = false;

		if (finalSystem[prop] || additionalProps.indexOf(prop) >= 0) {
			found = true;
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	}

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	return {
		base,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
