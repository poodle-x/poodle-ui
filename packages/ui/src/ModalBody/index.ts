import { ModalBody, ModalBodyProps } from "./ModalBody";
import { StandardThemeConfig } from "../theme";

export * from "./ModalBody";

export default ModalBody;

export type ModalBodyStyleKeys = "root";

export interface ModalBodyThemeConfig
	extends StandardThemeConfig<ModalBodyStyleKeys, ModalBodyProps> {}
