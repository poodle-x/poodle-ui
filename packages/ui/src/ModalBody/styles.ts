import { createCSSSystemStandard } from "../theme";
import { CSSSystem } from "../styled";
import { ModalBodyProps } from "./ModalBody";
import { ModalBodyStyleKeys } from "./";

export function Root(props: ModalBodyProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard<ModalBodyStyleKeys>({
		key: "root",
		config: theme?.ModalBody,
		props,
		base: [
			{
				minHeight: 200,
				py: "scale-4",
				px: "scale-6",
			},
		],
	});
}
