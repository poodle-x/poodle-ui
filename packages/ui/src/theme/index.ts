import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
