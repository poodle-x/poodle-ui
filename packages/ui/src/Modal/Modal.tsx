import React from "react";
import FocusTrap from "focus-trap-react";
import { Options as FocusTrapOptions } from "focus-trap";
import Portal, { PortalProps } from "../Portal";
import Box, { BoxProps } from "../Box";
import { StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../utils/useDefaultProps";
import { useClassNames } from "../styled";
import * as styles from "./styles";
import useAutoId from "../hooks/useAutoId";
import { ModalContext } from ".";

export interface LocalModalProps {
	children?: React.ReactNode;
	/**
	 * Open the modal when **`true`**.
	 */
	isOpen?: boolean;
	/**
	 * Callback called when the user tries to close the modal.
	 *
	 * It can be caused by click on outside the modal or press escape key.
	 */
	onRequestClose?: (cause: "outside" | "esc") => void;
	/**
	 * Custom props for inner element.
	 */
	innerProps?: BoxProps;
	/**
	 * Custom props for container element.
	 */
	containerProps?: BoxProps;
	/**
	 * Custom props for overlay element.
	 */
	overlayProps?: BoxProps;
	/**
	 * Custom props for root element.
	 */
	rootProps?: BoxProps;
	/**
	 * Custom props for modal element.
	 */
	modalProps?: BoxProps;
	/**
	 * Turn off focus trap.
	 */
	disableFocusTrap?: boolean;
	/**
	 * Custom props for **`<Portal />`**.
	 */
	portalProps?: Omit<PortalProps, "mount">;
	/**
	 * Add autoId
	 */
	withAutoId?: boolean;
	/**
	 * Pause or unpause focus trap
	 */
	pausedFocusTrap?: boolean;
	/**
	 * The elements will be used as the boundaries for the focus-trap,
	 * instead of the child.
	 */
	containerElementsFocusTrap?: Array<HTMLElement>;
	/**
	 * Pass any of the options available in focus-trap's createOptions.
	 */
	focusTrapOptions?: FocusTrapOptions;
}

export interface ModalProps
	extends BoxProps,
		StandardComponentProps,
		LocalModalProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.Modal?.defaultProps;
}

export const Modal: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<ModalProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, ModalProps>((_props, ref) => {
	const props = useDefaultProps<ModalProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		className,
		onRequestClose,
		innerProps,
		containerProps,
		modalProps,
		overlayProps,
		portalProps,
		isOpen,
		disableFocusTrap,
		withAutoId,
		pausedFocusTrap,
		containerElementsFocusTrap,
		focusTrapOptions,
		...otherProps
	} = props;

	const autoIdLabelledby = useAutoId(undefined, {
		base: "poodle-modal-label",
	});
	const autoIdDescribedby = useAutoId(undefined, {
		base: "poodle-modal-describe",
	});

	const mouseDownTarget = React.useRef<HTMLElement | null>(null);

	function handlerOverlayClick(e: React.MouseEvent<HTMLElement>) {
		if (e.target !== e.currentTarget) {
			return;
		}

		if (mouseDownTarget.current !== e.target) {
			return;
		}

		mouseDownTarget.current = null;

		if (onRequestClose) {
			onRequestClose("outside");
		}
	}

	function handlerOverlayMouseDown(e: React.MouseEvent<HTMLElement>) {
		mouseDownTarget.current = e.target as HTMLElement;
	}

	function handleKeyDown(e: React.KeyboardEvent<HTMLElement>) {
		if (e.key !== "Escape") {
			return;
		}

		e.stopPropagation();

		if (onRequestClose) {
			onRequestClose("esc");
		}
	}

	const classes = useClassNames({
		props,
		lists: {
			root: {
				classNames: ["poodle-modal", styles.Root, className],
			},
			overlay: {
				classNames: [
					"poodle-modal__overlay",
					styles.Overlay,
					overlayProps?.className,
				],
			},
			inner: {
				classNames: [
					"poodle-modal__inner",
					styles.Inner,
					innerProps?.className,
				],
			},
			container: {
				classNames: [
					"poodle-modal__container",
					styles.Container,
					containerProps?.className,
				],
			},
			modal: {
				classNames: [
					"poodle-modal__modal",
					styles.Modal,
					modalProps?.className,
				],
			},
		},
	});

	const renderChild = (
		<Box onKeyDown={handleKeyDown} {...otherProps} className={classes.root}>
			<Box aria-hidden={true} {...overlayProps} className={classes.overlay} />
			<Box
				onMouseDown={handlerOverlayMouseDown}
				onClick={handlerOverlayClick}
				{...innerProps}
				className={classes.inner}
			>
				<Box tabIndex={-1} {...containerProps} className={classes.container}>
					<Box
						role="dialog"
						aria-modal={true}
						aria-labelledby={withAutoId ? autoIdLabelledby : undefined}
						aria-describedby={withAutoId ? autoIdDescribedby : undefined}
						{...modalProps}
						className={classes.modal}
					>
						{children}
					</Box>
				</Box>
			</Box>
		</Box>
	);

	return (
		<ModalContext.Provider
			value={{
				autoIdDescribedby: withAutoId ? autoIdDescribedby : undefined,
				autoIdLabelledby: withAutoId ? autoIdLabelledby : undefined,
			}}
		>
			<Portal {...portalProps} ref={ref} mount={isOpen}>
				<FocusTrap
					containerElements={containerElementsFocusTrap}
					paused={pausedFocusTrap}
					active={!disableFocusTrap}
					focusTrapOptions={{
						allowOutsideClick: true,
						...focusTrapOptions,
					}}
				>
					{renderChild}
				</FocusTrap>
			</Portal>
		</ModalContext.Provider>
	);
});

Modal.displayName = "PoodleModal";
