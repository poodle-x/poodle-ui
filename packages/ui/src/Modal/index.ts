import React from "react";
import { Modal, ModalProps } from "./Modal";
import { StandardThemeConfig } from "../theme";

export * from "./Modal";

export default Modal;

export type ModalStyleKeys =
	| "root"
	| "overlay"
	| "inner"
	| "container"
	| "modal";

export interface ModalThemeConfig
	extends StandardThemeConfig<ModalStyleKeys, ModalProps> {}

export interface ModalContextValue {
	autoIdLabelledby?: string;
	autoIdDescribedby?: string;
}

export const ModalContext = React.createContext<ModalContextValue>({});
