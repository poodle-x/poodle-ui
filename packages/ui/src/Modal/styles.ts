import { getSpacing } from "@poodle/system";
import { createCSSSystemStandard } from "../theme";
import { CSSObjectSystem, CSSSystem } from "../styled";
import { ModalProps } from "./Modal";
import safeThemeUnitValue from "../utils/safeThemeValue";
import { ModalStyleKeys } from "./index";

export function Root(props: ModalProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard<ModalStyleKeys>({
		key: "root",
		config: theme?.Modal,
		props,
		base: [
			{
				position: "fixed",
				top: 0,
				left: 0,
				width: "100%",
				height: "100%",
				zIndex: "modal",
			},
		],
	});
}

export function Overlay(props: ModalProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard<ModalStyleKeys>({
		key: "overlay",
		config: theme?.Modal,
		props,
		base: [
			{
				position: "fixed",
				top: 0,
				left: 0,
				width: "100%",
				height: "100%",
				zIndex: -1,
				backgroundColor: "rgba(0, 0, 0, 0.25)",
			},
		],
	});
}

export function Inner(props: ModalProps): CSSSystem {
	const { theme } = props;

	return createCSSSystemStandard({
		key: "inner",
		config: theme?.Modal,
		props,
		base: [
			{
				display: "flex",
				alignItems: "center",
				justifyContent: "center",
				height: "100%",
				outline: "none",
			},
		],
	});
}

export function Container(props: ModalProps): CSSSystem {
	const { theme } = props;

	const topBottomMargin = getSpacing("scale-20")({ theme });

	return createCSSSystemStandard<ModalStyleKeys>({
		key: "container",
		config: theme?.Modal,
		props,
		base: [
			{
				maxHeight: `calc(100% - ${`${safeThemeUnitValue(topBottomMargin)}`})`,
				mx: "scale-4",
				my: "scale-10",
				width: "100%",
				maxWidth: "600px",
			},
		],
	});
}

export function Modal(props: ModalProps): CSSSystem {
	const { theme, colorStyle } = props;

	let colorStyles: CSSObjectSystem = {};

	switch (colorStyle) {
		case "primary": {
			colorStyles = {
				borderTop: "4px solid",
				borderColor: "primary.base",
			};
			break;
		}
		case "negative": {
			colorStyles = {
				borderTop: "4px solid",
				borderColor: "negative.base",
			};
			break;
		}
		case "positive": {
			colorStyles = {
				borderTop: "4px solid",
				borderColor: "positive.base",
			};
			break;
		}
		case "warn": {
			colorStyles = {
				borderTop: "4px solid",
				borderColor: "warn.base",
			};
			break;
		}
	}

	return createCSSSystemStandard({
		key: "modal",
		config: theme?.Modal,
		props,
		base: [
			{
				width: "100%",
				maxHeight: "100%",
				borderRadius: "4px",
				backgroundColor: "modal.base",
				boxShadow: "lv1",
			},
			colorStyles,
		],
	});
}
