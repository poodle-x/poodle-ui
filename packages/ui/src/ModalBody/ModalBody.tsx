import React, { useContext } from "react";
import Box, { BoxProps } from "../Box";
import { ModalContext } from "../Modal";
import { getCSSSystemBoxProps } from "../styled/CSSSystem";
import { StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../hooks/useDefaultProps/useDefaultProps";
import * as styles from "./styles";

export interface LocalModalBodyProps {
	children?: React.ReactNode;
}

export interface ModalBodyProps
	extends BoxProps,
		StandardComponentProps,
		LocalModalBodyProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.ModalBody?.defaultProps;
}

export const ModalBody: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<ModalBodyProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, ModalBodyProps>((_props, ref) => {
	const { props, isLocalTheme } = useDefaultProps<ModalBodyProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});
	const { autoIdDescribedby } = useContext(ModalContext);

	const { children, ...otherProps } = props;

	return (
		<Box
			id={autoIdDescribedby}
			{...otherProps}
			{...getCSSSystemBoxProps({
				isRoot: true,
				isLocalTheme,
				componentProps: props,
				fnCSSSystem: styles.Root,
				baseClassName: ["poodle-modal-body"],
			})}
			ref={ref}
		>
			{children}
		</Box>
	);
});

ModalBody.displayName = "PoodleModalBody";
