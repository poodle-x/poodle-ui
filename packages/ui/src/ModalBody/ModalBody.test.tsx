import React from "react";
import { ModalContext } from "../Modal";
import { ThemeConfig } from "../theme";
import { render } from "../utils/test";
import ModalBody from "./";

describe("<ModalBody />", () => {
	it("should render ModalBody correctly", () => {
		const { container } = render(<ModalBody>ModalBody</ModalBody>);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render ModalBody with auto id", () => {
		const { container } = render(
			<ModalContext.Provider
				value={{
					autoIdDescribedby: "id",
				}}
			>
				<ModalBody>ModalBody</ModalBody>
			</ModalContext.Provider>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("ref should work", () => {
		const ref = React.createRef<HTMLDivElement>();

		render(<ModalBody ref={ref}>ModalBody</ModalBody>);
		expect(ref.current).toMatchSnapshot();
	});

	it("should render Box props correctly", () => {
		const { container } = render(
			<ModalBody id="id" mt="20px">
				ModalBody
			</ModalBody>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			ModalBody: {
				styles: {
					root: {
						mt: 10,
						letterSpacing: 250,
					},
				},
				variants: {
					custom: {
						root: {
							top: 0,
							textIndent: 1,
						},
					},
				},
			},
		};

		it("should extend styles in global", () => {
			const { container } = render(
				<ModalBody variant="custom">ModalBody</ModalBody>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend styles in local", () => {
			const { container } = render(
				<ModalBody variant="custom" themeExtend={themeTest}>
					ModalBody
				</ModalBody>
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should render with default props correctly", () => {
			const { container } = render(<ModalBody>ModalBody</ModalBody>, {
				theme: {
					ModalBody: {
						defaultProps: {
							id: "id",
							className: "hello",
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles", () => {
			const { container } = render(<ModalBody>ModalBody</ModalBody>, {
				theme: {
					ModalBody: {
						overrides: {
							root: {
								mt: 50,
							},
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
