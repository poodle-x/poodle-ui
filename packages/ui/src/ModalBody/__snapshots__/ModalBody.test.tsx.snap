// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalBody /> ref should work 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
}

<div
  class="poodle-box poodle-modal-body c0"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> should render Box props correctly 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
  margin-top: 20px;
}

<div
  class="poodle-box poodle-modal-body c0"
  id="id"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> should render ModalBody correctly 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
}

<div
  class="poodle-box poodle-modal-body c0"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> should render ModalBody with auto id 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
}

<div
  class="poodle-box poodle-modal-body c0"
  id="id"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> theming should extend styles in global 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<div
  class="poodle-box poodle-modal-body c0"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> theming should extend styles in local 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
  margin-top: 10px;
  letter-spacing: 250px;
  top: 0;
  text-indent: 1px;
}

<div
  class="poodle-box poodle-modal-body c0"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> theming should override styles 1`] = `
.c0 {
  margin-top: 50px;
}

<div
  class="poodle-box poodle-modal-body c0"
>
  ModalBody
</div>
`;

exports[`<ModalBody /> theming should render with default props correctly 1`] = `
.c0 {
  min-height: 200px;
  padding-top: 16px;
  padding-bottom: 16px;
  padding-left: 24px;
  padding-right: 24px;
}

<div
  class="poodle-box poodle-modal-body hello c0"
  id="id"
>
  ModalBody
</div>
`;
